### PR TITLE
[AIRFLOW-6056] Allow EmrAddStepsOperator to accept job_flow_name as alternative to job_flow_id

### PR DESCRIPTION
**Pull request by aviemzur**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-XXX
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.
  - In case you are proposing a fundamental code change, you need to create an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)).
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:
Allow EmrAddStepsOperator to accept job_flow_name as alternative to job_flow_id
This is useful for when you have a persistent EMR cluster for testing or for when you have repeated scheduled jobs with a small interval between them.

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### CODE REVIEW

**Comment by OmerJog** [comment]
@aviemzur can you fix error?
`airflow/contrib/hooks/emr_hook.py:48:111: E501 line too long (115 > 110 characters)
`

**Comment by aviemzur** [comment]
> @aviemzur can you fix error?
> `airflow/contrib/hooks/emr_hook.py:48:111: E501 line too long (115 > 110 characters) `

Yup, just pushed a fix

**Comment by aviemzur** [comment]
R: @kaxil

File: airflow/contrib/hooks/emr_hook.py [hunk 1]
```diff
@@ -36,6 +36,26 @@ def get_conn(self):
...
+            ClusterStates=[
+                'RUNNING', 'WAITING'
+            ]
+        )
+
+        matching_clusters = list(
+            filter(lambda cluster: cluster['Name'] == emr_cluster_name, response['Clusters'])
+        )
+
+        if len(matching_clusters) > 0:
```
Comment (ashb) [inline airflow/contrib/hooks/emr_hook.py]: If there is more than a single cluster of the given name that should probably be an error.

File: airflow/contrib/hooks/emr_hook.py [hunk 2]
```diff
@@ -36,6 +36,26 @@ def get_conn(self):
         self.conn = self.get_client_type('emr', self.region_name)
         return self.conn
 
+    def get_cluster_id_by_name(self, emr_cluster_name):
+        conn = self.get_conn()
+
+        response = conn.list_clusters(
+            ClusterStates=[
+                'RUNNING', 'WAITING'
```
Comment (ashb) [inline airflow/contrib/hooks/emr_hook.py]: I would suggest making the list of valid states to be a parameter -- in the past I have done something like this and included `'BOOTSTRAPPING'` in my list.

File: airflow/contrib/operators/emr_add_steps_operator.py [hunk 3]
```diff
@@ -28,6 +28,9 @@ class EmrAddStepsOperator(BaseOperator):
 
     :param job_flow_id: id of the JobFlow to add steps to. (templated)
     :type job_flow_id: str
+    :param job_flow_name: name of the JobFlow to add steps to (alternative to passing job_flow_id. will
+        search for id of first running/waiting JobFlow with matching name). (templated)
```
Comment (ashb) [inline airflow/contrib/operators/emr_add_steps_operator.py]: This is not templated yet - you need to add it to `template_fields` list.

File: airflow/contrib/operators/emr_add_steps_operator.py [hunk 4]
```diff
@@ -40,18 +43,26 @@ class EmrAddStepsOperator(BaseOperator):
...
+            job_flow_id=None,
+            job_flow_name=None,
             aws_conn_id='s3_default',
             steps=None,
             *args, **kwargs):
         super().__init__(*args, **kwargs)
         steps = steps or []
-        self.job_flow_id = job_flow_id
         self.aws_conn_id = aws_conn_id
+        self.emr_hook = EmrHook(aws_conn_id=self.aws_conn_id)
```
Comment (ashb) [inline airflow/contrib/operators/emr_add_steps_operator.py]: Please don't do this -- don't make connections/requests inside `__init__`, it should only be done in `execute()`

File: airflow/contrib/hooks/emr_hook.py [hunk 5]
```diff
@@ -36,6 +36,26 @@ def get_conn(self):
         self.conn = self.get_client_type('emr', self.region_name)
```
Comment (ashb) [inline airflow/contrib/hooks/emr_hook.py]:
```suggestion
        if not self.conn:
            self.conn = self.get_client_type('emr', self.region_name)
```

(and do `self.conn = None` inside `__init__`)

File: tests/contrib/operators/test_emr_add_steps_operator.py [hunk 6]
```diff
@@ -93,14 +109,22 @@ def test_render_template(self):
     def test_execute_returns_step_id(self):
         self.emr_client_mock.add_job_flow_steps.return_value = ADD_STEPS_SUCCESS_RETURN
 
-        # Mock out the emr_client creator
-        emr_session_mock = MagicMock()
-        emr_session_mock.client.return_value = self.emr_client_mock
-        self.boto3_session_mock = MagicMock(return_value=emr_session_mock)
```
Comment (ashb) [inline tests/contrib/operators/test_emr_add_steps_operator.py:99]: Why did you remove this block?

**Review by ashb — CHANGES_REQUESTED** [review]

File: tests/contrib/operators/test_emr_add_steps_operator.py (hunk 6 above)
Comment (aviemzur) [reply tests/contrib/operators/test_emr_add_steps_operator.py:99]: Wasn't removed but moved to `setUp` since both the existing test case and the new test case use the mock.

**Comment by aviemzur** [comment]
@ashb Added commit with the changes you requested.

Changed the operator to push `job_flow_id` to `xcom` so it can be used to pass to other operators for example `EmrStepSensor`.

File: airflow/contrib/operators/emr_add_steps_operator.py [hunk 7]
```diff
@@ -28,33 +28,57 @@ class EmrAddStepsOperator(BaseOperator):
...
-            aws_conn_id='aws_default',
-            steps=None,
-            *args, **kwargs):
+        self,
+        job_flow_id=None,
+        job_flow_name=None,
+        cluster_states=None,
+        aws_conn_id='s3_default',
+        steps=None,
+        *args, **kwargs):
```
Comment (ashb) [inline airflow/contrib/operators/emr_add_steps_operator.py]: This formatting looks odd - please could you keep in indented as it was?

File: airflow/contrib/operators/emr_add_steps_operator.py [hunk 8]
```diff
@@ -28,33 +28,57 @@ class EmrAddStepsOperator(BaseOperator):
...
+        if kwargs.get('xcom_push') is not None:
+            raise AirflowException("'xcom_push' was deprecated, use 'do_xcom_push' instead")
         super().__init__(*args, **kwargs)
+        if not job_flow_id and not job_flow_name:
+            raise AirflowException('Either job_flow_id or job_flow_name must be specified.')
```
Comment (ashb) [inline airflow/contrib/operators/emr_add_steps_operator.py]: We should also check that both are not specified, right?

File: tests/contrib/operators/test_emr_add_steps_operator.py [hunk 9]
```diff
@@ -93,13 +101,30 @@ def test_render_template(self):
...
+                aws_conn_id='aws_default',
+                dag=DAG('test_dag_id', default_args=self.args)
+            )
+
+            operator.execute(self.mock_context)
+
+            ti = self.mock_context['ti']
+            job_flow_id = ti.xcom_push.mock_calls[0][2]['value']
+
+            self.assertEqual(job_flow_id, expected_job_flow_id)
```
Comment (ashb) [inline tests/contrib/operators/test_emr_add_steps_operator.py]: This would be more clear as:

```python
    ti.xcom_push.assert_any_call(key='job_flow_id', value=expected_job_flow_id)
```

**Review by ashb — CHANGES_REQUESTED** [review]

**Comment by aviemzur** [comment]
@ashb Added commit with the changes you requested + lint fixes [Changes after CR 2](https://github.com/apache/airflow/pull/6655/commits/abae66a9c4a8968aea7718052370c2571122e2ee)

File: airflow/contrib/operators/emr_add_steps_operator.py [hunk 10]
```diff
@@ -28,33 +28,58 @@ class EmrAddStepsOperator(BaseOperator):
...
 
     @apply_defaults
     def __init__(
             self,
-            job_flow_id,
-            aws_conn_id='aws_default',
+            job_flow_id=None,
+            job_flow_name=None,
+            cluster_states=None,
+            aws_conn_id='s3_default',
```
Comment (ashb) [inline airflow/contrib/operators/emr_add_steps_operator.py]: You've changed the default connection ID. You shouldn't.

File: airflow/contrib/operators/emr_add_steps_operator.py [hunk 11]
```diff
@@ -28,33 +28,58 @@ class EmrAddStepsOperator(BaseOperator):
...
             steps=None,
             *args, **kwargs):
+        if kwargs.get('xcom_push') is not None:
+            raise AirflowException("'xcom_push' was deprecated, use 'do_xcom_push' instead")
+        if (job_flow_id is None) + (job_flow_name is None) != 1:
```
Comment (ashb) [inline airflow/contrib/operators/emr_add_steps_operator.py]: This is a slightly odd way of expressing this. Either use logcial XOR:

```suggestion
        if (job_flow_id is None) ^ (job_flow_name is None):
```

Or do it as

```suggestion
        if job_flow_id is None or job_flow_name is None or (job_flow_name and job_flow_id):
```

**Comment by ashb** [comment]
You also need to fix the failing static checks. You should be able to run those locally.

File: airflow/contrib/operators/emr_add_steps_operator.py (hunk 10 above)
Comment (aviemzur) [reply airflow/contrib/operators/emr_add_steps_operator.py]: Made an error while rebasing, will fix.

**Comment by aviemzur** [comment]
@ashb yeah had a problem with missing packages, now I run the static checks locally.

File: airflow/contrib/operators/emr_add_steps_operator.py (hunk 11 above)
Comment (aviemzur) [reply airflow/contrib/operators/emr_add_steps_operator.py]: XOR doesn't fulfill the conditional needed here, I can change it to not + xor, will this look better?
```suggestion
        if not ((job_flow_id is None) ^ (job_flow_name is None)):
```

Comment (ashb) [reply airflow/contrib/operators/emr_add_steps_operator.py]: Oh yes :)

Which ever of this or the more verbose form you find clearer to understand.

**Comment by aviemzur** [comment]
@ashb Added commit with the changes you requested + lint fixes [Changes after CR 3](https://github.com/apache/airflow/pull/6655/commits/2249a3dd5f54477d9e8960963626d48e3538252f)

**Review by ashb — APPROVED** [review]

**Comment by gorros** [comment]
@aviemzur  This is a great feature. Thanks for contribution. Currently I have couple of long term running clusters and i used to do same what you do. I was listing clusters and filtering by name. But that was resulted in throttling exception from AWS API when number of such dags was growing. Then I started to persist id of currently active cluster. Before submitting a step I was describing the cluster and making sure it is still running. This reduced number of throttling exceptions but still there were some. Then I introduced exception handling and [exponential back off ](https://stackoverflow.com/questions/27438273/exponential-backoff-time-sleep-with-random-randint0-1000-1000). This final step eliminated throttling issue. Also does your implementation take care of listing cluster with pagination?

**Comment by aviemzur** [comment]
@gorros Exponential back off was introduced to sensors in Airflow's master already.
Regarding the API throttling we are aware of this and I intend to open an Airflow issue to have a base AWS sensor which will not fail if such an error happens.

**Comment by gorros** [comment]
@aviemzur but I think back off is needed for hook too.

**Comment by aviemzur** [comment]
Gotcha,
Then there should also be an Airflow issue to have a base AWS hook which uses exponential backoff.
